tests/v3/connection_state_test.c

```diff
@@ -2424,9 +2485,328 @@ static int s_test_mqtt_connection_unsub_timeout_fn(struct aws_allocator *allocat
+        .client_id = aws_byte_cursor_from_c_str("client1234"),
+        .host_name = aws_byte_cursor_from_c_str(state_test_data->endpoint.address),
+        .socket_options = &state_test_data->socket_options,
+        .on_connection_complete = aws_test311_on_connection_complete_fn,
+        .ping_timeout_ms = DEFAULT_TEST_PING_TIMEOUT_MS,
+        .keep_alive_time_secs = 16960, /* basically stop automatically sending PINGREQ */
+    };
+
+    struct aws_byte_cursor pub_topic = aws_byte_cursor_from_c_str("/test/topic");
+
+    ASSERT_SUCCESS(aws_mqtt_client_connection_connect(state_test_data->mqtt_connection, &connection_options));
+    aws_test311_wait_for_connection_to_complete(state_test_data);
+
+    /* Disable the auto ACK packets sent by the server, which blocks the requests to complete */
+    mqtt_mock_server_disable_auto_ack(state_test_data->mock_server);
+
+    aws_mutex_lock(&state_test_data->lock);
+    state_test_data->expected_ops_completed = 1;
+    aws_mutex_unlock(&state_test_data->lock);
+
+    /* unsubscribe to the first topic */
+    uint16_t unsub_packet_id = aws_mqtt_client_connection_311_unsubscribe(
+        state_test_data->mqtt_connection->impl,
+        &pub_topic,
+        aws_test311_on_op_complete,
+        state_test_data,
+        aws_timestamp_convert(3, AWS_TIMESTAMP_SECS, AWS_TIMESTAMP_NANOS, NULL));
+    ASSERT_TRUE(unsub_packet_id > 0);
+
+    /* unsubscribe should complete after the timeout */
+    aws_test311_wait_for_ops_completed(state_test_data);
+    /* Check that the unsubscribe has been completed with a timeout error */
+    ASSERT_UINT_EQUALS(state_test_data->op_complete_error, AWS_ERROR_MQTT_TIMEOUT);
+    ASSERT_SUCCESS(aws_mqtt_client_connection_disconnect(
+        state_test_data->mqtt_connection, aws_test311_on_disconnect_fn, state_test_data));
+    aws_test311_wait_for_disconnect_to_complete(state_test_data);
+
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE_FIXTURE(
+    mqtt_connection_unsubscribe_timeout_override,
+    s_setup_mqtt_server_fn,
+    s_test_mqtt_connection_unsubscribe_timeout_override_fn,
+    s_clean_up_mqtt_server_fn,
+    &test_data)
+
+/**
+ * Test that connection is healthy, user set the timeout for request, and timeout happens and the subscribe failed.
+ */
+static int s_test_mqtt_connection_subscribe_single_timeout_fn(struct aws_allocator *allocator, void *ctx) {
+    (void)allocator;
+    struct mqtt_connection_state_test *state_test_data = ctx;
+
+    struct aws_mqtt_connection_options connection_options = {
+        .user_data = state_test_data,
+        .clean_session = false,
+        .client_id = aws_byte_cursor_from_c_str("client1234"),
+        .host_name = aws_byte_cursor_from_c_str(state_test_data->endpoint.address),
+        .socket_options = &state_test_data->socket_options,
+        .on_connection_complete = aws_test311_on_connection_complete_fn,
+        .ping_timeout_ms = DEFAULT_TEST_PING_TIMEOUT_MS,
+        .protocol_operation_timeout_ms = 3000,
```

Comment:
It looks like we're using the same value for the regular timeout tests and overridden timeout tests (3 seconds). Let's use different values (shorter for override) to make sure we're using the override value when examining test results. Also, might be useful to set the `protocol_operation_timeout_ms` in the override tests to make sure we're overriding the provided value explicitly. Not setting it defaults the regular value to `UINT64_MAX` which is ignored so any timeout occurring probably indicates this test is successful.

Reply:
I'm not bothered by this.  Elapsed time comparisons in tests have consisently blown up in our faces when containers and CPU throttling on cloud CI (and emulated ARM builds) get thrown into the mix.